### PR TITLE
#123 Implement memory footprint mode

### LINKED ISSUE #123: Implement memory footprint mode
Implement a new mode and possibly a new operator for computing the memory footprint that has overloads for no serializable but sizable elements like `std::function<>` or pointers (just count the point size and then follow it).

### PR DESCRIPTION
**Pull request by cz4rs**
fixes #123

### CODE REVIEW

**Comment by cz4rs** [comment]
note:
pushing some code here, but it's definitely still work-in-progress (requires improved naming (!) and docs, reduced code duplication, independence from `serialize` / separate operator, implementation for other containers etc.)

**Comment by cz4rs** [comment]
@lifflander I have removed the custom dispatcher and used existing `serialize` code when possible, can you have a look?

- standard containers, fundamental types, enums seem to be working fine, code for `std::tuple`, `std::function` and other necessary types will land tomorrow
- the tests are failing currently because of how raw pointers are handled (currently just counting the pointer size and following it, but that doesn't work for C-style arrays and C strings) - any idea how to proceed with this?
- I have updated https://github.com/DARMA-tasking/vt/pull/1013 to have an example on the `vt` side, if that looks ok I will start working on components one by one to see if any issues come out
- how much testing is required for when `detector` is not available? (currently the CI only tests with `detector` present)

File: src/checkpoint/container/array_serialize.h
Comment (PhilMiller) [inline src/checkpoint/container/array_serialize.h]: I think this is double-counting, since the implementation is required to contain the `T[N]` as a member, not point to it. So, I think the `countBytes` can just be omitted

File: src/checkpoint/container/chrono_serialize.h
Comment (PhilMiller) [inline src/checkpoint/container/chrono_serialize.h]: Since this can only compute a footprint, it should probably assert that the serializer is in that mode, rather than silently failing

File: src/checkpoint/container/queue_serialize.h
Comment (PhilMiller) [inline src/checkpoint/container/queue_serialize.h]: These implementations are destructive, which is quite undesirable.

File: src/checkpoint/container/raw_ptr_serialize.h
Comment (PhilMiller) [inline src/checkpoint/container/raw_ptr_serialize.h]: Again, since this will only work for footprinting, it should assert `s.isFootprinting()`

File: src/checkpoint/container/shared_ptr_serialize.h
Comment (PhilMiller) [inline src/checkpoint/container/shared_ptr_serialize.h]: Ditto about assertion

File: src/checkpoint/container/string_serialize.h
Comment (PhilMiller) [inline src/checkpoint/container/string_serialize.h:75]: This too potentially runs into the small-array optimization, but I think it's *not* a problem here, because `std::string` has a *fixed* small array size, so the overcount is tightly limited

**Comment by PhilMiller** [comment]
Overall, I like this.

We'll want to figure out some systematic way to identify whether a class needs special footprint handling, and whether we've considered the need for special handling in each class for which we have a serializer.

Since we're handling container classes of various sorts in checkpoint, the serializers for `Kokkos` types need to be considered, too. I'm a little concerned that we don't correctly handle the special case of `Kokkos::View<Kokkos::View<T*>>` - the nested views will take up space even if empty, and need to be counted.

File: src/checkpoint/container/queue_serialize.h
Comment (cz4rs) [reply src/checkpoint/container/queue_serialize.h]: I couldn't find a better (concise) way to iterate over queue's elements, so I went with passing by value - `pop` is called on the copy. This is not too elegant and potentially slow, but the original object remains untouched.

File: src/checkpoint/checkpoint_api.impl.h
Comment (cz4rs) [inline src/checkpoint/checkpoint_api.impl.h]: at this stage (of implementation) `getMemoryFootprint` sometimes returns smaller value than a simple `sizeof`
(this can be valid calculation because of padding / layout)

note to self: change to `std::max(footprint, sizeof(T))` before merging

File: src/checkpoint/container/raw_ptr_serialize.h
Comment (lifflander) [inline src/checkpoint/container/raw_ptr_serialize.h]: I think we need to bifurcate the dispatch for non-serializable, but footprintable types. Or cases where we want footprinting to be handled quite differently than regular serialization  We could do that statically, with an enable_if on the `SerializerT`. See comment at the bottom.

File: src/checkpoint/container/raw_ptr_serialize.h
Comment (lifflander) [inline src/checkpoint/container/raw_ptr_serialize.h]: Same here.

File: src/checkpoint/container/chrono_serialize.h
Comment (lifflander) [reply src/checkpoint/container/chrono_serialize.h]: For this one, can we add a real serializer also here? I don't think chrono should be that difficult to serialize

File: src/checkpoint/container/queue_serialize.h
Comment (lifflander) [reply src/checkpoint/container/queue_serialize.h]: It's quite unfortunate that this container (queue) doesn't have access to the underlying data structure.

**Comment by lifflander** [comment]
@cz4rs @PhilMiller I'd like to include enable_if's on footprinting-only serializers, so that they can't accidentally be used by other serializers. This also makes it clear why the code doesn't compile if some footprint specific member (like `s.countBytes`) is called and then fails to work with Packer/Unpacker/Sizer.

```C++
template <
  typename SerializerT,
  typename T,
  typename = std::enable_if_t<std::is_same<SerializerT, checkpoint::Footprinter>::value>
>
void serialize(SerializerT& s, std::queue<T> queue) {

}
```

If we want to make this code more general (assuming different footprinters might be used), we might want to check for a type trait instead of the exact type `Footprinter`.

File: src/checkpoint/container/raw_ptr_serialize.h
Comment (cz4rs) [reply src/checkpoint/container/raw_ptr_serialize.h]: enforcing Footprinting mode with `std::enable_if_t`

**Comment by cz4rs** [comment]
> I'd like to include enable_if's on footprinting-only serializers, so that they can't accidentally be used by other serializers. This also makes it clear why the code doesn't compile if some footprint specific member (like `s.countBytes`) is called and then fails to work with Packer/Unpacker/Sizer.

I have implemented `enable_if`'s in a helper function called by `serialize` and using anything else than `getMemoryFootprint` fails during compilation as expected. Putting `enable_if` directly in `serialize` was causing problems with proper detection in `dispatch_serializer_nonbyte.h`.

File: src/checkpoint/container/shared_ptr_serialize.h
Comment (cz4rs) [reply src/checkpoint/container/shared_ptr_serialize.h]: `std_enable_if_t` added

File: src/checkpoint/container/raw_ptr_serialize.h
Comment (cz4rs) [inline src/checkpoint/container/raw_ptr_serialize.h:56]: handling arrays correctly could be possible using `std::is_array` or
```cpp
template<typename SerializerT, typename T, size_t N> 
void serialize(T (&t)[N])
```

I poked around it a little bit, but currently arrays decay into pointers in the middle of call stack and I have left the warning here

File: src/checkpoint/container/array_serialize.h
Comment (cz4rs) [reply src/checkpoint/container/array_serialize.h]: removed the redundant `countBytes` call

File: src/checkpoint/container/chrono_serialize.h
Comment (cz4rs) [reply src/checkpoint/container/chrono_serialize.h]: serializer added

File: src/checkpoint/container/queue_serialize.h
Comment (PhilMiller) [inline src/checkpoint/container/queue_serialize.h]: Per discussion on today's call, we prefer to just count `q.size() * sizeof(T)`, and sacrifice some accuracy to avoid the potential pitfall of copying and iterating over every element.

Comment (PhilMiller) [reply src/checkpoint/container/queue_serialize.h]: That would shift us back to taking the queue arguments by `const&`

File: src/checkpoint/container/raw_ptr_serialize.h
Comment (PhilMiller) [inline src/checkpoint/container/raw_ptr_serialize.h]: On some platforms, I believe `struct FILE;` is left as an incomplete definition, to avoid exposing the C library's implementation. I think this would fail to compile in those settings, since it's ultimately going to attempt `sizeof()` on an incomplete type.

File: src/checkpoint/container/string_serialize.h
Comment (PhilMiller) [inline src/checkpoint/container/string_serialize.h]: Typo - should be "its"

File: src/checkpoint/container/raw_ptr_serialize.h
Comment (cz4rs) [reply src/checkpoint/container/raw_ptr_serialize.h]: changed to size the pointer only

File: src/checkpoint/container/string_serialize.h
Comment (cz4rs) [reply src/checkpoint/container/string_serialize.h]: fixed

File: src/checkpoint/container/queue_serialize.h
Comment (cz4rs) [reply src/checkpoint/container/queue_serialize.h]: done, the inaccuracy will be bigger when the underlying container is `std::vector` (default for priority queue), but there's no access to capacity in that case

**Comment by cz4rs** [comment]
@PhilMiller @lifflander if this looks ok for you, I would like to get it merged to have better CI feedback in https://github.com/DARMA-tasking/vt/pull/1013

(I would be adding support for kokkos types and ironing out any discovered bugs in a subsequent PR)

**Review by PhilMiller — APPROVED** [review]

File: src/checkpoint/container/queue_serialize.h
Comment (PhilMiller) [reply src/checkpoint/container/queue_serialize.h]: I realize that for the specific cases in the runtime of these containers, we can handle footprinting specially, by keeping a running measure of the footprint of the contents, to add to the footprint explicitly.

Which means, the footprint interface needs to be able to accept "add arbitrary byte count"

**Comment by cz4rs** [comment]
@lifflander @PhilMiller (thanks for the review!)
regarding the merge permissions / protected branches: please take a look at the screenshot below, I cannot merge this pull request myself
- the build passed fine
- review is approved "by reviewer with write access"
- the "not authorized" comment links to https://docs.github.com/en/github/administering-a-repository/about-protected-branches

I don't have a fresh example from `vt` repository, but it looked pretty similar there (we discussed this some time ago on DARMA technical meeting)

image

**Comment by lifflander** [comment]
> @lifflander @PhilMiller (thanks for the review!)
> regarding the merge permissions / protected branches: please take a look at the screenshot below, I cannot merge this pull request myself

I believe that I have fixed this now for this repository.

**Pull request merged**
